src/network/network_crypto_internal.h

```diff
@@ -0,0 +1,317 @@
+			const X25519SecretKey &our_secret_key, const X25519PublicKey &our_public_key, std::string_view extra_payload);
+};
+
+/**
+ * Base for handlers using a X25519 key exchange to perform authentication.
+ *
+ * In general this works as follows:
+ * 1) the client and server have or generate a secret and public X25519 key.
+ * 2) the X25519 key exchange is performed at both the client and server, with their own secret key and their peer's public key.
+ * 3) a pair of derived keys is created by BLAKE2b-hashing the following into 64 bytes, in this particular order:
+ *    - the shared secret from the key exchange;
+ *    - the public key of the server;
+ *    - the public key of the client;
+ *    - optional extra payload, e.g. a password in the case of PAKE.
+ *    The first of the pair of derived keys is usually used to encrypt client-to-server communcation, and the second of the pair
+ *    is usually used to encrypt server-to-client communication.
+ * 4) a XChaCha20-Poly1305 (authenticated) encryption is performed using:
+ *    - the first of the pair of derived keys as encryption key;
+ *    - a 24 byte nonce;
+ *    - the public key of the client as additional authenticated data.
+ *    - a 8 byte random number as content/message.
+ *
+ * The server initiates the request by sending its public key and a 24 byte nonce that is randomly generated. Normally the side
+ * that sends the encrypted data sends the nonce in their packet, which would the the client on our case. However, there are
+ * many implementations of clients due to the admin-protocol where this is used, and we cannot guarantee that they generate a
+ * good enough nonce. As such the server sends one instead. The server will create a new set of keys for each session.
+ *
+ * The client receives the request, performs the key exchange, generates the derived keys and then encrypts the message. This
+ * message must contain some content, so it has to be filled with a four random bytes. Once the message has been encrypted, the
+ * client sends their public key, the encrypted message and the message authentication code (MAC) to the server in a response.
+ *
+ * The server receives the response, performs the key exchange, generates the derived keys, decrypts the message and validates the
+ * message authentication code, and finally the message. It is up to the sub class to perform the final authentication checks.
+ */
+class X25519AuthenticationHandler {
+private:
+	X25519SecretKey our_secret_key; ///< The secret key used by us.
+	X25519PublicKey our_public_key; ///< The public key used by us.
+	X25519Nonce nonce; ///< The nonce to prevent replay attacks.
+	X25519DerivedKeys derived_keys; ///< Keys derived from the authentication process.
+
+protected:
+	X25519PublicKey peer_public_key; ///< The public key used by our peer.
+
+	X25519AuthenticationHandler(const X25519SecretKey &secret_key = X25519SecretKey::CreateRandom());
```

Comment:
This line has been .. weirding me out. Mostly, I really did not know you could call a function as a default value like this in C++ :P But it also took me a bit of time to find this statement.

But the other thing that kinda bothers me about this line, is who is the owner of the memory?

`CreateRandom()` returns a `X25519SecretKey`. This ctor stores a pointer to that object. But .. who then owns the object? How is the memory still alive? When will it be released?

If I would be to guess, the reason this works is because `X25519SecretKey` is in fact a `std::array`. But it raises a lot of eyebrows.

Shouldn't we just move this out, and not have a default constructor here? I am aware that this means you need to store the `secret_key` in the client-handler on the server-side. But at least that makes more clear who owns the memory?

It might also be that I am just missing something here :)

Reply:
Welcome to the magic world of C++ object lifetime.

What happens is that C++ makes a temporary object with the created secret key, then runs the constructor with that and afterwards C++ cleans up that temporary object. Since it's an object on the stack and not the heap, there is no memory management that needs to be done.

See https://godbolt.org/z/oEr3breKn for this magic in action with some output on the (de)constructors etc.

Reply:
The trick that happens here, is that although you receive a pointer, you make a copy of the object directly after. So indeed, a lifetime of just the ctor call is sufficient. Still very scary code.

Either way, I am still not sure if it is actually a good idea to have this here. This is really hidden, and really took me a while to find where the server generates his public/secret.

Reply:
Guess I would prefer the more explicit route of:

```
        if (this->authentication_handler == nullptr) {
-               this->authentication_handler = NetworkAuthenticationServerHandler::Create(&_password_provider, &_authorized_key_handler);
+               /* Create a new secret key for every connection. */
+               X25519SecretKey secret_key = X25519SecretKey::CreateRandom()
+               this->authentication_handler = NetworkAuthenticationServerHandler::Create(secret_key&, &_password_provider, &_authorized_key_handler);
        }
```

That way it is also documented what the behaviour is, and that it is expected.

Reply:
The biggest problem I have with that proposal is having to expose the internal crypto guts in the API. 
So basically anything that includes `network/core/core.h` (indirectly) will also get the internal crypto guts included, because the internal details of some secret key needed to be exposed.

If you want to avoid that, you'd need to pass a `std::string` there and if it's empty or invalid regenerate a secret key.

So, could we just add comment to `NetworkAuthenticationServerHandler::Create` that a new public/secret key pair is generated for each authentication method?
By the way, for the client only the authorized key one uses the provided public/secret keypair. The others also generate a new public/secret pair for each method.